peps/pep-0800.rst

```diff
@@ -0,0 +1,355 @@
+
+For example, in this program::
+
+    def f(x: bool) -> None:
+        if isinstance(x, str):
+            print("It's both!")
+
+both pyright and mypy (with ``--warn-unreachable``), two popular type checkers, will warn that the body of the
+``if`` block is unreachable, because if ``x`` is a ``bool``, it cannot also be a ``str``.
+
+Reachability is complicated in Python by the presence of multiple inheritance. If instead of ``bool`` and ``str``,
+we use two user-defined classes, mypy and pyright do not show any warnings::
+
+    class A: pass
+    class B: pass
+
+    def f(x: A):
+        if isinstance(x, B):
+            print("It's both!")
+
+This is correct, because a class that inherits from both ``A`` and ``B`` could exist.
+
+We see a divergence between type checkers in another case, where we use ``int`` and ``str``::
+
+    def f(x: int):
+        if isinstance(x, str):
+            print("It's both!")
+
+For this code, pyright shows no errors but mypy will claim that the branch is unreachable. Mypy is technically correct
+here: CPython does not allow a class to inherit from both ``int`` and ``str``, so the branch is unreachable.
+However, the information necessary to determine that these base classes are incompatible is not currently available in
+the type system. Mypy, in fact, uses a heuristic based on the presence of incompatible methods; this heuristic works
+reasonably well in practice, especially for built-in types, but it is
+`incorrect in general <https://github.com/python/mypy/issues/19377>`__.
+
+The experimental ``ty`` type checker uses a third approach that aligns more closely with the runtime behavior of Python:
+it recognizes certain classes as "solid bases" that restrict multiple inheritance. Broadly speaking, every class must
+inherit from at most one unique solid base, and if there is no unique solid base, the class cannot exist; we'll provide a more
+precise definition below. However, ty's approach relies on hardcoded knowledge of particular built-in types.
+
+This PEP proposes an extension to the type system that makes it possible to express when multiple inheritance is not
+allowed at runtime: an ``@solid_base`` decorator that marks classes as "solid bases".
+This gives type checkers a more precise understanding of reachability, and helps in several concrete areas.
+
+Invalid class definitions
+-------------------------
+
+The following class definition raises an error at runtime, because ``int`` and ``str`` are distinct solid bases::
+
+    class C(int, str): pass
+
+Without knowledge of solid bases, type checkers are not currently able to detect the reason why this class
+definition is invalid, though they may detect that if this class were to exist, some of its methods would be incompatible.
+(When it sees this class definition, mypy will point at incompatible definitions of ``__add__`` and several other
+methods.)
```

Comment:
this still doesn't explain the concept of "incompatible definitions" -- if we wanted to, I think it would look something like this. But again, this might be too long outside of a footnote.

```suggestion
definition is invalid, though they may detect that if this class were to exist, some of its methods would
implicitly override methods in its superclass in a way that would be inconsistent with
the Liskov Substitution Principle.
(For example, when mypy sees this class definition, mypy will point at "incompatible definitions" of ``__add__``:
`int` will appear earlier in the MRO, so ``C.__add__`` will be inherited from `int.__add__`;
but this constitutes an implicit -- and Liskov-incompatible -- override of ``str.__add__``
due to the MRO of ``C``.)
```

Reply:
I think I'll write an appendix discussing this a bit more.

Reply:
Nice, the appendix is great!